File: core/backup_manager.py
import os
import time
import logging
import threading
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional

from core.utilities import create_backup

class BackupManager:
    """Manages server backup operations including scheduling and cleanup."""
    
    _instance = None
    _logger = None
    _initialized = False

    def __new__(cls, server, pause_event=None):
        if cls._instance is None:
            cls._instance = super(BackupManager, cls).__new__(cls)
        return cls._instance

    def __init__(self, server, pause_event=None):
        """Initialize backup manager."""
        # Skip if already initialized
        if BackupManager._initialized:
            return
            
        # Setup logger first
        if BackupManager._logger is None:
            BackupManager._logger = self._setup_logger()
        self.logger = BackupManager._logger
            
        self.server = server
        self.backup_thread = None
        self.stop_flag = threading.Event()
        self.pause_event = pause_event if pause_event else threading.Event()
        self.pause_event.set()  # Start in non-paused state
        
        # Initialize backup timing
        self.next_backup = None
        
        # Mark as initialized but don't start automatically
        BackupManager._initialized = True
        self.update_from_config(auto_start=False)

    def _setup_logger(self):
        """Setup dedicated logger for backup manager"""
        logger = logging.getLogger('backup_manager')
        
        # Only add handlers if they don't exist
        if not logger.handlers:
            formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
            
            # Console handler
            console_handler = logging.StreamHandler()
            console_handler.setFormatter(formatter)
            logger.addHandler(console_handler)
            
            # File handler
            try:
                log_dir = Path("logs")
                log_dir.mkdir(exist_ok=True)
                
                file_handler = logging.FileHandler(log_dir / "backup_manager.log")
                file_handler.setFormatter(formatter)
                logger.addHandler(file_handler)
            except Exception as e:
                print(f"Failed to setup file logging: {e}")
            
            logger.setLevel(logging.INFO)
            
        # Prevent propagation to root logger
        logger.propagate = False
        
        return logger

    def _calculate_next_backup(self) -> None:
        """Calculate the next backup time based on current configuration"""
        try:
            auto_backup_config = self.server.config.get('auto_backup', {})
            if auto_backup_config.get('enabled', False):
                interval_minutes = auto_backup_config.get('interval_minutes', 360)
                self.next_backup = datetime.now() + timedelta(minutes=interval_minutes)
                self.logger.info(f"Next backup scheduled for: {self.next_backup}")
            else:
                self.next_backup = None
                self.logger.info("Auto-backup is disabled")
        except Exception as e:
            self.logger.error(f"Error calculating next backup time: {e}")
            self.next_backup = None

    def update_from_config(self, auto_start=True):
        """Update backup manager state based on current configuration"""
        try:
            auto_backup_config = self.server.config.get('auto_backup', {})
            
            if auto_backup_config.get('enabled', False):
                self.logger.info("Auto-backup is enabled in config")
                
                if not self.next_backup:
                    interval = auto_backup_config.get('interval_minutes', 360)
                    self.next_backup = datetime.now() + timedelta(minutes=interval)
                    self.logger.info(f"Next backup scheduled for: {self.next_backup}")
                
                if auto_start and not (self.backup_thread and self.backup_thread.is_alive()):
                    self.start()
            else:
                self.logger.info("Auto-backup is disabled in config")
                if self.backup_thread and self.backup_thread.is_alive():
                    self.stop()
                    
        except Exception as e:
            self.logger.error(f"Error updating from config: {e}")

    def start(self) -> bool:
        """Start the backup manager thread"""
        if self.backup_thread and self.backup_thread.is_alive():
            self.logger.info("Backup manager already running")
            return False

        try:
            # Only start if server is running
            if not self.server.is_running():
                self.logger.info("Server not running, backup manager will not start")
                return False

            self.stop_flag.clear()
            self.backup_thread = threading.Thread(target=self._backup_loop, name="BackupThread")
            self.backup_thread.daemon = True
            self.backup_thread.start()
            self.logger.info("Backup manager started")
            return True
        except Exception as e:
            self.logger.error(f"Failed to start backup manager: {e}")
            return False

    def stop(self) -> bool:
        """Stop the backup manager thread"""
        try:
            if self.backup_thread and self.backup_thread.is_alive():
                self.logger.info("Stopping backup manager")
                self.stop_flag.set()
                self.backup_thread.join(timeout=30)
                
                if self.backup_thread.is_alive():
                    self.logger.warning("Backup thread did not stop gracefully")
                else:
                    self.logger.info("Backup manager stopped successfully")
                    
                self.backup_thread = None
            return True
        except Exception as e:
            self.logger.error(f"Error stopping backup manager: {e}")
            return False

    def _backup_loop(self) -> None:
        """Main backup loop"""
        self.logger.info("Backup loop started")
        last_cleanup_check = datetime.now()
        
        while not self.stop_flag.is_set():
            try:
                # Wait for any pause to be lifted
                self.pause_event.wait()
                
                # Check if server is still running
                if not self.server.is_running():
                    self.logger.info("Server not running, stopping backup manager")
                    self.stop()
                    break

                current_time = datetime.now()
                
                # Check if backup is due
                if self.next_backup and current_time >= self.next_backup:
                    self.logger.info("Starting scheduled backup")
                    
                    # Check and cleanup old backups daily
                    if current_time - last_cleanup_check > timedelta(days=1):
                        if self._should_cleanup():
                            self._cleanup_old_backups()
                        last_cleanup_check = current_time
                    
                    # Perform backup
                    if self.server.backup():
                        self.logger.info("Automated backup completed successfully")
                        # Recalculate next backup time
                        self._calculate_next_backup()
                    else:
                        self.logger.error("Automated backup failed")
                        # Retry in 5 minutes on failure
                        self.next_backup = current_time + timedelta(minutes=5)
                
                # Sleep for 1 minute
                for _ in range(6):  # Check stop flag every 10 seconds
                    if self.stop_flag.is_set():
                        break
                    time.sleep(10)
                    
            except Exception as e:
                self.logger.error(f"Error in backup loop: {e}")
                time.sleep(60)  # Wait before retrying after error

    def _should_cleanup(self) -> bool:
        """Check if cleanup should be performed"""
        try:
            backup_dir = Path(self.server.config["backup_dir"])
            if not backup_dir.exists():
                return False
            
            backup_files = list(backup_dir.glob("enshrouded_backup_*.zip"))
            if not backup_files:
                return False
                
            auto_backup_config = self.server.config.get('auto_backup', {})
            keep_days = auto_backup_config.get('keep_days', 7)
            
            oldest_allowed = datetime.now() - timedelta(days=keep_days)
            
            for backup_file in backup_files:
                try:
                    date_str = backup_file.stem.split('_')[2:4]
                    file_date = datetime.strptime('_'.join(date_str), '%Y%m%d_%H%M%S')
                    if file_date < oldest_allowed:
                        return True
                except (ValueError, IndexError):
                    continue
                    
            return False
        except Exception as e:
            self.logger.error(f"Error checking cleanup status: {e}")
            return False

    def _cleanup_old_backups(self) -> None:
        """Clean up old backup files"""
        try:
            auto_backup_config = self.server.config.get('auto_backup', {})
            keep_days = auto_backup_config.get('keep_days', 7)
            backup_dir = Path(self.server.config["backup_dir"])
            
            if not backup_dir.exists():
                return
                
            cutoff_date = datetime.now() - timedelta(days=keep_days)
            deleted_count = 0
            failed_count = 0
            
            for backup_file in backup_dir.glob("enshrouded_backup_*.zip"):
                try:
                    date_str = backup_file.stem.split('_')[2:4]
                    file_date = datetime.strptime('_'.join(date_str), '%Y%m%d_%H%M%S')
                    
                    if file_date < cutoff_date:
                        backup_file.unlink()
                        deleted_count += 1
                        self.logger.info(f"Deleted old backup: {backup_file.name}")
                except (ValueError, IndexError) as e:
                    self.logger.warning(f"Could not parse date from backup filename {backup_file.name}: {e}")
                    failed_count += 1
                except Exception as e:
                    self.logger.error(f"Error deleting backup {backup_file.name}: {e}")
                    failed_count += 1
            
            if deleted_count > 0 or failed_count > 0:
                self.logger.info(f"Cleanup completed: {deleted_count} files deleted, {failed_count} failed")
                    
        except Exception as e:
            self.logger.error(f"Error during backup cleanup: {e}")

    def get_next_backup_time(self) -> Optional[datetime]:
        """Get the next scheduled backup time"""
        return self.next_backup

    def force_backup(self) -> bool:
        """Force an immediate backup"""
        try:
            if self.server.backup():
                self.logger.info("Forced backup completed successfully")
                # Recalculate next scheduled backup time
                self._calculate_next_backup()
                return True
            else:
                self.logger.error("Forced backup failed")
                return False
        except Exception as e:
            self.logger.error(f"Error during forced backup: {e}")
            return False

    def shutdown(self) -> None:
        """Perform cleanup on shutdown"""
        try:
            auto_backup_config = self.server.config.get('auto_backup', {})
            if auto_backup_config.get('backup_on_shutdown', True):
                self.logger.info("Performing shutdown backup")
                self.server.backup()
            self.stop()
        except Exception as e:
            self.logger.error(f"Error during shutdown: {e}")

File: Core/BackupManager.cs
using System.IO.Compression;
using EnshroudedServerManager.Models;

namespace EnshroudedServerManager.Core;

public class BackupManager
{
    private readonly ServerManager _server;
    private ServerConfig Config => _server.Config;

    private CancellationTokenSource? _cts;
    private Task? _loopTask;

    public DateTime? NextBackup { get; private set; }

    public BackupManager(ServerManager server)
    {
        _server = server;
    }

    private int _scheduledIntervalMinutes;

    public void Start()
    {
        // Same rule as RestartManager: a loop with a cancelled token is as good
        // as stopped — replace it (happens when a restart cycles this manager).
        if (_loopTask is { IsCompleted: false } && _cts is { IsCancellationRequested: false })
        {
            AppLogger.Info("Backup manager already running.");
            return;
        }

        ScheduleNext();
        _cts = new CancellationTokenSource();
        var token = _cts.Token;
        _loopTask = Task.Run(() => RunLoopAsync(token));
        AppLogger.Info("Backup manager started.");
    }

    public void Stop()
    {
        _cts?.Cancel();
        NextBackup = null;
        AppLogger.Info("Backup manager stop requested.");
    }

    public void UpdateFromConfig()
    {
        if (Config.AutoBackup.Enabled)
        {
            if (_loopTask is null or { IsCompleted: true } || _cts is null or { IsCancellationRequested: true })
            {
                if (_server.IsRunning())
                    Start();
            }
            else if (Config.AutoBackup.IntervalMinutes != _scheduledIntervalMinutes)
            {
                ScheduleNext(); // interval changed — recompute; otherwise keep the current schedule
            }
        }
        else
        {
            Stop();
        }
    }

    /// <param name="performBackup">
    /// False when the caller already ran the shutdown backup (StopAsync does it
    /// after stopping the server) or the server is still running — zipping a
    /// live savegame risks a corrupt archive.
    /// </param>
    public async Task ShutdownAsync(bool performBackup = true)
    {
        if (performBackup && Config.AutoBackup.BackupOnShutdown)
        {
            AppLogger.Info("Performing shutdown backup...");
            await PerformBackupAsync();
        }
        Stop();
        if (_loopTask != null)
            await Task.WhenAny(_loopTask, Task.Delay(10_000));
    }

    private void ScheduleNext()
    {
        if (Config.AutoBackup.Enabled)
        {
            _scheduledIntervalMinutes = Config.AutoBackup.IntervalMinutes;
            NextBackup = DateTime.Now.AddMinutes(_scheduledIntervalMinutes);
            AppLogger.Info($"Next backup scheduled: {NextBackup:yyyy-MM-dd HH:mm:ss}");
        }
        else
        {
            NextBackup = null;
        }
    }

    private async Task RunLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                if (!_server.IsRunning())
                {
                    AppLogger.Info("Server stopped — backup manager exiting.");
                    NextBackup = null;
                    break;
                }

                var now = DateTime.Now;

                if (NextBackup.HasValue && now >= NextBackup.Value)
                {
                    bool ok = await PerformBackupAsync();
                    NextBackup = ok
                        ? DateTime.Now.AddMinutes(Config.AutoBackup.IntervalMinutes)
                        : DateTime.Now.AddMinutes(5); // retry sooner on failure

                    AppLogger.Info($"Next backup scheduled: {NextBackup:yyyy-MM-dd HH:mm:ss}");
                }

                await Task.Delay(TimeSpan.FromSeconds(30), ct);
            }
            catch (OperationCanceledException)
            {
                break;
            }
            catch (Exception ex)
            {
                AppLogger.Error($"Error in backup loop: {ex.Message}");
                await Task.Delay(TimeSpan.FromMinutes(1), ct).ConfigureAwait(false);
            }
        }

        AppLogger.Info("Backup manager stopped.");
    }

    public async Task<bool> PerformBackupAsync()
    {
        try
        {
            var backupDir = Config.BackupDir;
            Directory.CreateDirectory(backupDir);

            var saveDir = Path.Combine(Config.ServerDir, "savegame");
            if (!Directory.Exists(saveDir))
            {
                AppLogger.Error($"Save directory not found: {saveDir}");
                return false;
            }

            var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
            var backupPath = Path.Combine(backupDir, $"enshrouded_backup_{timestamp}.zip");

            AppLogger.Info($"Creating backup: {backupPath}");

            await Task.Run(() =>
            {
                if (File.Exists(backupPath))
                    File.Delete(backupPath);

                ZipFile.CreateFromDirectory(saveDir, backupPath, CompressionLevel.Optimal, false);
            });

            AppLogger.Info($"Backup created successfully: {backupPath}");
            CleanupOldBackups();
            return true;
        }
        catch (Exception ex)
        {
            AppLogger.Error($"Backup failed: {ex.Message}");
            return false;
        }
    }

    private void CleanupOldBackups()
    {
        try
        {
            var backupDir = Config.BackupDir;
            if (!Directory.Exists(backupDir)) return;

            var cutoff = DateTime.Now.AddDays(-Config.AutoBackup.KeepDays);
            int deleted = 0, failed = 0;

            foreach (var file in Directory.GetFiles(backupDir, "enshrouded_backup_*.zip"))
            {
                try
                {
                    // Parse date from filename: enshrouded_backup_YYYYMMDD_HHMMSS.zip
                    var stem = Path.GetFileNameWithoutExtension(file);
                    var parts = stem.Split('_');
                    if (parts.Length >= 4)
                    {
                        var dateStr = $"{parts[2]}_{parts[3]}";
                        if (DateTime.TryParseExact(dateStr, "yyyyMMdd_HHmmss",
                            null, System.Globalization.DateTimeStyles.None, out var fileDate))
                        {
                            if (fileDate < cutoff)
                            {
                                File.Delete(file);
                                deleted++;
                                AppLogger.Info($"Deleted old backup: {Path.GetFileName(file)}");
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    AppLogger.Warning($"Could not process backup file {Path.GetFileName(file)}: {ex.Message}");
                    failed++;
                }
            }

            if (deleted > 0 || failed > 0)
                AppLogger.Info($"Cleanup complete: {deleted} deleted, {failed} failed.");
        }
        catch (Exception ex)
        {
            AppLogger.Error($"Error during backup cleanup: {ex.Message}");
        }
    }
}
